src/super_gradients/training/models/segmentation_models/segformer.py

```diff
@@ -0,0 +1,516 @@
+            x = x.reshape(b_size, h, w, -1).permute(0, 3, 1, 2)
+
+            features.append(x)
+
+        return features
+
+
+class MLP(nn.Module):
+    def __init__(self, dim, embed_dim):
+        """
+        A single Linear layer, with shape pre-processing
+        :param dim: input dimension
+        :param embed_dim: output dimension
+        """
+
+        super().__init__()
+
+        self.proj = nn.Linear(dim, embed_dim)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = x.flatten(2).transpose(1, 2)
+        x = self.proj(x)
+
+        return x
+
+
+class LinearFuse(nn.Module):
+    def __init__(self, in_channels: int, out_channels: int):
+        """
+        A linear fusion block (conv + bn + relu) (https://arxiv.org/pdf/2105.15203.pdf)
+        :param in_channels: number of input channels
+        :param out_channels: number of output channels
+        """
+
+        super().__init__()
+
+        self.conv = nn.Conv2d(in_channels=in_channels, out_channels=out_channels, kernel_size=1, bias=False)
+        self.bn = nn.BatchNorm2d(out_channels)
+        self.relu = nn.ReLU(inplace=True)
+
+    def forward(self, x):
+        return self.relu(self.bn(self.conv(x)))
+
+
+class SegFormerHead(nn.Module):
+    def __init__(self, encoder_dims: list, embed_dim: int, num_classes: int):
+        """
+        SegFormer decoder head (https://arxiv.org/pdf/2105.15203.pdf)
+        :param encoder_dims: list of encoder embedding dimensions
+        :param embed_dim: unified embedding dimension
+        :param num_classes: number of predicted classes
+        """
+        super().__init__()
+
+        self.linear_layers = []
+        for idx, dim in enumerate(encoder_dims):
+            self.linear_layers.append(MLP(dim, embed_dim))
+            self.add_module(f"linear_c{idx + 1}", self.linear_layers[idx])
+
+        self.linear_fuse = LinearFuse(in_channels=embed_dim * len(encoder_dims), out_channels=embed_dim)
+        self.linear_pred = nn.Conv2d(in_channels=embed_dim, out_channels=num_classes, kernel_size=1)
+
+        self.dropout = nn.Dropout2d(0.1)
+
+    def forward(self, features: List[torch.Tensor]) -> torch.Tensor:
+        b, _, h, w = features[0].shape
+
+        out_lst = [self.linear_layers[0](features[0]).permute(0, 2, 1).reshape(b, -1, *features[0].shape[-2:])]
+
+        for i, feature in enumerate(features[1:]):
+            out = self.linear_layers[i + 1](feature).permute(0, 2, 1).reshape(b, -1, *feature.shape[-2:])
+            out = F.interpolate(out, size=(h, w), mode="bilinear", align_corners=False)
+            out_lst.append(out)
+
+        out = self.linear_fuse(torch.cat(out_lst[::-1], dim=1))
+        out = self.linear_pred(self.dropout(out))
+
+        return out
+
+
+# TODO: add support for aux heads? (not in original impl) (currently not using)
+class SegFormer(SegmentationModule):
+    def __init__(
+        self,
+        num_classes: int,
+        encoder_embed_dims: list,
```

Comment:
please use full typing on all of the params:
`List[int]` or `List[float]` instead of `list`